File: Javascript/Linked List/21.合并两个有序链表.js
/*
 * @lc app=leetcode.cn id=21 lang=javascript
 *
 * [21] 合并两个有序链表
 */

// @lc code=start
/**
 * Definition for singly-linked list.
 * function ListNode(val) {
 *     this.val = val;
 *     this.next = null;
 * }
 */
/**
 * @param {ListNode} l1
 * @param {ListNode} l2
 * @return {ListNode}
 */
// var mergeTwoLists = function(l1, l2) {
//     const merge = (l1, l2) => {
//         if (l1 == null) return l2;
//         if (l2 == null) return l1;

//         if (l1.val > l2.val) {
//             l2.next = merge(l1, l2.next);
//             return l2;
//         }else {
//             l1.next = merge(l1.next, l2);
//             return l1;
//         }

//     }

//     return merge(l1, l2);
// };
var mergeTwoLists = function(l1, l2) {
    if (l1 == null) return l2;
    if (l2 == null) return l1;

    let p = dummyHead = new ListNode();

    let p1 = l1, p2 =l2;

    while(p1 && p2) {
        if (p1.val > p2.val) {
            p.next = p2;
            p = p.next;
            p2 = p2.next;
        } else {
            p.next = p1;
            p = p.next;
            p1 = p1.next;
        }
    }

    if (p1) {
        p.next = p1;
    } else {
        p.next = p2;
    }

    return dummyHead.next;
}
// @lc code=end

File: Go/LinkedList/21-合并两个有序链表.go
package LinkedList

// l1头部额外加一个，差位比较
func mergeTwoLists(l1 *ListNode, l2 *ListNode) *ListNode {
	l1 = &ListNode{
		Val:  0,
		Next: l1,
	}
	tmp := &ListNode{}
	head := l1
	for l1.Next != nil && l2 != nil {
		if l2.Val > l1.Next.Val {
			l1 = l1.Next
		} else {
			tmp = l2.Next
			l2.Next = l1.Next
			l1.Next = l2
			l2 = tmp
		}
	}
	if l2 != nil {
		// l1跑到尾部
		for l1.Next != nil {
			l1 = l1.Next
		}
		// 把l2剩余的续上去
		for l2 != nil {
			l1.Next = l2
			l1 = l1.Next
			l2 = l2.Next
		}
	}
	return head.Next
}